### ISSUE TITLE
device_id is not set properly when _infer_torch_distributed_launch_init is called

### ISSUE THREAD

**Issue by tangyuq** [pytorch/fairseq#3385]
## 🐛 Bug

When lauching multi-node distributed training with `python -m torch.distributed.launch`, all processes are setting device_id to 0. 

After adding the following to _infer_torch_distributed_launch_init ( https://github.com/pytorch/fairseq/blob/master/fairseq/distributed/utils.py#L80), 
   cfg.device_id = cfg.local_rank
the issue is fixed

### To Reproduce

rank=0 or 1
master=ip_address of the master node 
python -m torch.distributed.launch \
    --nproc_per_node=8  \
    --nnodes=2 \
    --node_rank=$rank \
    --master_addr="$master" \
    --master_port=12346  \
    $(which fairseq-train) \
    "<multi30k_dataset>/multi30k/bin" \
    --max-update 20000 \
    --save-dir <checkpoint_path>/mult30k_debug \
    --source-lang de --target-lang en \
    --arch transformer_iwslt_de_en --share-decoder-input-output-embed \
    --optimizer adam --adam-betas '(0.9, 0.98)' --clip-norm 0.0 \
    --fp16 \
    --lr 5e-4 --lr-scheduler inverse_sqrt --warmup-updates 4000 \
    --dropout 0.3 --weight-decay 0.0001 \
    --criterion label_smoothed_cross_entropy --label-smoothing 0.1 \
    --max-tokens 1000 \
    --log-interval 1 \
    --maximize-best-checkpoint-metric 
  

### Expected behavior

Only GPU 0 has load in each node; and the training stucks at line: https://github.com/pytorch/fairseq/blob/master/fairseq/trainer.py#L90

### Environment

 - fairseq Version (e.g., 1.0 or master): master 
 - PyTorch Version (e.g., 1.0): 1.7.1
 - OS (e.g., Linux): Linux
 - How you installed fairseq (`pip`, source): source
 - Build command you used (if compiling from source):
 - Python version: 3.6.5
 - CUDA/cuDNN version: 
 - GPU models and configuration:
 - Any other relevant information:
